### ISSUE TITLE
Multicast packets received but IP_MULTICAST_ALL unset

### ISSUE THREAD

**Issue by sim-** [acassen/keepalived#952]
Hello! Per the discussion at the end of issue #931, this is a separate issue to discuss/track that issue.

Keepalived_vrrp[26989]: Kernel/system configuration issue causing multicast packets to be received but IP_MULTICAST_ALL unset

I added some debugging to the kernel net/ipv4/igmp.c ip_mc_sf_allow() to see when and why it would return 1, which is why such packets are delivered to sockets. I wrote a standalone receiver that just calls socket() and setsockopt(...,IP_MULTICAST_ALL,...), and it appears to function as what keepalived is seeing.

Long story short, the simple answer here is that I am using a unicast_peer block, and this causes non-multicast packets to arrive. The ip_mc_sf_allow() kernel function returns 1 right at the first test because the loc_addr is not a multicast address.

So, maybe the answer here is to add some sort of filter to the write-only sockets, or to just leave the reading in place and sink the packets without a warning.

**Comment by pqarmitage** [comment]
Prior to commit 6fb5980 keepalived wasn't reading on the write sockets, and for sufficiently recent kernels there were no packets building up in the socket receive queue (as shown by `netstat -anp | grep keepalived`. On some older kernels (certainly 3.6 through 3.10) packets were being queued to the receive queue.

In order to stop the receive queue building up when using an older kernel, I added the receive on the write sockets. Since more recent kernels weren't having any build up on the receive queue, I assumed that no packets were being queued to it, and so I added the `IP_MULTICAST_ALL unset` warning, expecting it only to be reported on older kernels, but that appears not to be the case.

What I would like to understand is what packets are being received on the send socket, and why aren't we seeing receive queues building on the socket when we don't try to receive on the socket.

**Comment by pqarmitage** [comment]
I've added some code to keepalived and what I am seeing is that it is IPv6 packets that are being received, and fortunately only on IPv6 send sockets. If I run keepalived with no IPv6 instances I do not get the `IP_MULTICAST_ALL unset` warning.

So I suspect we now need to find if there is an IPv6 equivalent of IP_MULTICAST_ALL and use that for Ipv6 multicast sockets.

**Comment by pqarmitage** [comment]
Commit 77d947f makes keepalived only report the `IP_MULTICAST_ALL unset` for IPv4, which for stop the error message occurring at all.

The message will continue to be reported for kernels which still queue IPv4 packets on the send socket (e.g. 3.6 kernels).

**Comment by pqarmitage** [comment]
I think the question now is whether there is a way to stop the IPv6 multicast packets being queued to the send socket. If not, then I think either a kernel patch is produced to provide the equivalent of IP_MULTICAST_ALL for IPv6, or we close this issue.

**Comment by sim-** [comment]
But you would also see it for unicast IPv4 VRRP packets, right? Try setting unicast_src_ip to the local interface IP and unicast_peer to the remote host where you trace the UDP sockets receiving packets. With a simple raw socket receiver (e.g.: http://0x.ca/sim/tmp/udprecv.c ), it picks up the VRRP packets for me. This is what is causing them the write sockets to receive stuff in my case (IPv6 is still turned off here). Actually, that reminds me to test IPv6 again -- last time I tested it, IPv6 neighbor discovery was causing problems with vmac. :)

Anyway, there may something that is causing leaky multicast filtering on those kernels which is separate from this, but I don't see any kernel mechanism other than maybe SOL_SOCKET, SO_ATTACH_FILTER to truly filter everything.

**Comment by pqarmitage** [comment]
@sim- Yes, I am seeing it for unicast IPv4 packets.

I have added the following function, using SO_ATTACH_FILTER as you suggested, and `if_setsockopt_no_receive()` is called for all send sockets It appears to be successfully stopping any packets being queued to the socket:
```
int
if_setsockopt_no_receive(int *sd)
{
        int ret;
        struct sock_filter bpfcode[1] = {
                {0x06, 0, 0, 0},        /* ret #0 */
        };      
        struct sock_fprog bpf = {1, bpfcode};

        if (*sd < 0)
                return -1;

        ret = setsockopt(*sd, SOL_SOCKET, SO_ATTACH_FILTER, &bpf, sizeof(bpf));
        if (ret < 0) {
                log_message(LOG_INFO, "Can't set SO_ATTACH_FILTER option. errno=%d (%m)", errno);
                close(*sd);
                *sd = -1;
        }

        return *sd;
}
```

Does this seem reasonable to you?

**Comment by acassen** [comment]
nice

**Comment by acassen** [comment]
There is certainly a root cause in kernel code that need time to spend on. Anyway cBPF/eBPF is certainly the best future proof work around.
